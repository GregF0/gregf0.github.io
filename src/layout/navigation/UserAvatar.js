import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import styled, { css } from 'styled-components'
import useMediaQuery from '@material-ui/core/useMediaQuery'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import MatAvatar from '@material-ui/core/Avatar'
import { startSignOut } from '../../data/auth/actions'
import { setSettings } from '../../data/settings/actions'
import { initialState } from '../../data/settings/reducer'

export const UserAvatar = () => {
  const { name, photo } = useSelector((state) => state.auth)
  const [anchorEl, setAnchorEl] = useState(null)
  const dispatch = useDispatch()

  const signOut = () => {
    dispatch(startSignOut())
    dispatch(setSettings(initialState))
    handleClose()
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = () => {
    setAnchorEl(null)
  }

  const sidenav = +useMediaQuery('(min-width:600px) and (min-height:500px)')

  return (
    <>
      <Avatar sidenav={sidenav} alt={name} src={photo} onClick={openMenu} />
      <Menu
        id="user-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={handleClose}
      >
        <MenuItem onClick={signOut}>Sign Out</MenuItem>
      </Menu>
    </>
  )
}

export const Avatar = styled(MatAvatar)`
  width: 30px;
  height: 30px;
  cursor: pointer;

  ${({ sidenav }) =>
    sidenav &&
    css`
      width: 35px;
      height: 35px;
    `}
`
