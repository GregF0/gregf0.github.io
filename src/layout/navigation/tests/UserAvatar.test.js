import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import Menu from '@material-ui/core/Menu'
import { UserAvatar, Avatar } from '../UserAvatar'

describe('<UserAvatar />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<UserAvatar />)
  }

  const dispatchMocked = jest.fn()
  jest.spyOn(redux, 'useDispatch').mockImplementation(() => dispatchMocked)

  const createStore = () => {
    const store = {
      auth: {
        name: 'Alex',
        photo: 'https://via.placeholder.com/100',
      },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <UserAvatar /> correctly', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })

  test('should render user image with correct alt attribute', () => {
    const { name, photo } = createStore().auth

    expect(createWrapper().find(Avatar).props()).toMatchObject({
      src: photo,
      alt: name,
    })
  })

  test('menu should be closed initially', () => {
    createStore()
    expect(createWrapper().find(Menu).props().open).toBe(false)
  })

  test('should open menu on avatar click', () => {
    createStore()
    const wrapper = createWrapper()

    wrapper.find(Avatar).simulate('click', { currentTarget: {} })
    expect(wrapper.find(Menu).props().open).toBe(true)
  })
})
