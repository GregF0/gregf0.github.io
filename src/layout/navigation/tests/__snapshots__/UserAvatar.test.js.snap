// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserAvatar /> should render <UserAvatar /> correctly 1`] = `
<Fragment>
  <Styled(WithStyles(ForwardRef(Avatar)))
    alt="Alex"
    onClick={[Function]}
    sidenav={0}
    src="https://via.placeholder.com/100"
  />
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    id="user-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      Sign Out
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</Fragment>
`;
